[Merged by Bors] - feat: ring with discrete PrimeSpectrum

alreadydone opened the pull request:
Prove two characterizations of commutative semirings R whose prime spectra have discrete topology:

+ R has finitely many prime ideals, and every prime ideal is maximal;

+ R has finitely many maximal ideals, the intersection of which (= Jacobson radical) is contained in (and therefore equal to) the nilradical.

From the first characterization it is easy to show the spectrum of an Artinian ring has discrete topology, but Artinian doesn't currently import PrimeSpectrum.

Also show that if R is a commutative ring and Spec R has discrete topology, then the canonical RingHom from R to the product of localizations at maximal ideals (which is always injective) is an isomorphism. This is not a characterization, because it is satisfied by every local ring. (But it's a characterization if we replace "maximal" by "prime".)

For this purpose we prove that, if a basic open D(f) in Spec R is a singleton {p}, then localization away from f is the same as localization at p. A key fact used is that if an ideal in a commutative semiring is disjoint from a submonoid, then there exists a maximal such ideal containing the ideal, and any maximal such ideal is prime.

Also upgrade `exists_idempotent_basicOpen_eq_of_isClopen` to `existsUnique`.

---

Add Stacks tags to work done by @erdOne in #14372 and #18394.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

erdOne commented on Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:
```diff
@@ -211,6 +211,31 @@ instance compactSpace : CompactSpace (PrimeSpectrum R) := by
+finitely many maximal ideals and their intersection is contained in the nilradical. -/
+theorem discreteTopology_iff_finite_isMaximal_and_sInf_le_nilradical :
+    letI s := {I : Ideal R | I.IsMaximal}
+    DiscreteTopology (PrimeSpectrum R) ↔ Finite s ∧ sInf s ≤ nilradical R :=
```
Should this be in terms of jacobson radical?

alreadydone replied:
This holds for (comm.) semirings while `Ideal.jacobson` is only defined for rings at the moment. Also it's convenient to use `sInf s` since we already have `s` around.

erdOne commented on Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:
```diff
@@ -890,24 +955,71 @@ lemma PrimeSpectrum.basicOpen_eq_zeroLocus_of_isIdempotentElem
+    basicOpen (isIdempotentElemEquivIsClopen (R := R).symm s).1 = s.1 :=
+  congr_arg (·.1) (isIdempotentElemEquivIsClopen.apply_symm_apply s)
+
+lemma IsLocalization.away_quotient_of_isIdempotentElem {e : R} (he : IsIdempotentElem e) :
+    IsLocalization.Away e (R ⧸ Ideal.span {1 - e}) :=
+  away_of_isIdempotentElem he Ideal.mk_ker Quotient.mk_surjective
```
I don't think this should be here.

erdOne commented on Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:
```diff
@@ -890,24 +955,71 @@ lemma PrimeSpectrum.basicOpen_eq_zeroLocus_of_isIdempotentElem
+@[stacks 00JA
+"See also `PrimeSpectrum.discreteTopology_iff_finite_isMaximal_and_sInf_le_nilradical`."]
+def RingHom.toLocalizationIsMaximalEquiv [DiscreteTopology (PrimeSpectrum R)] : R ≃+*
+    ∀ I : {I : Ideal R | I.IsMaximal}, haveI : I.1.IsMaximal := I.2; Localization.AtPrime I.1 :=
```
```suggestion
    Π I : {I : Ideal R | I.IsMaximal}, haveI : I.1.IsMaximal := I.2; Localization.AtPrime I.1 :=
```

alreadydone replied:
mathport translated Π to ∀, but it can still be used if it reduces confusion, I guess ...

erdOne commented on Mathlib/RingTheory/PrimeSpectrum.lean:
```diff
@@ -83,6 +83,13 @@ instance [Subsingleton R] : IsEmpty (PrimeSpectrum R) :=
 
 variable (R S)
 
+/-- The prime spectrum is in bijection with the set of prime ideals. -/
+def equivSetOfIsPrime : PrimeSpectrum R ≃ {I : Ideal R | I.IsPrime} where
```
```suggestion
/-- The prime spectrum is in bijection with the set of prime ideals. -/
@[simps]
def equivSetOfIsPrime : PrimeSpectrum R ≃ {I : Ideal R // I.IsPrime} where
```

alreadydone replied:
I usually prefer to use coercion from Set; they're defeq but not sure syntactically. Anyway it doesn't matter a lot.

alreadydone replied:
Looks like this broke the build, because I'm using the set of maximal ideals is contained in the set of primes; let me change it back.

alreadydone commented on Mathlib/RingTheory/PrimeSpectrum.lean:
```diff
@@ -83,6 +83,14 @@ instance [Subsingleton R] : IsEmpty (PrimeSpectrum R) :=
 
+/-- The prime spectrum is in bijection with the set of prime ideals. -/
+@[simps]
+def equivSetOfIsPrime : PrimeSpectrum R ≃ {I : Ideal R // I.IsPrime} where
```
```suggestion
def equivSetOfIsPrime : PrimeSpectrum R ≃ {I : Ideal R | I.IsPrime} where
```

erdOne reviewed (commented):
I forgot to submit the review. I see that you've already fixed some?

alreadydone commented on Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:
```diff
@@ -951,24 +1016,67 @@ lemma PrimeSpectrum.basicOpen_eq_zeroLocus_of_isIdempotentElem
+      SetLike.ext' (congr_arg (·.1) eq)), fun s ↦
+        have ⟨e, he, h⟩ := exists_idempotent_basicOpen_eq_of_isClopen s.2
+        ⟨⟨e, he⟩, Subtype.ext h.symm⟩⟩
+
```
Is
```
adding projection PrimeSpectrum.isIdempotentElemEquivIsClopen_apply_coe:
      ∀ {R : Type u} [inst : CommRing R] (e : ↑{e | IsIdempotentElem e}),
        ↑(PrimeSpectrum.isIdempotentElemEquivIsClopen e) = (PrimeSpectrum.zeroLocus {↑e})ᶜ 
```
what you want? (produced by @[simps!])

erdOne replied:
Yes but it should be `= basicOpen` and `coe` should be a prefix

erdOne commented on Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:
```diff
@@ -212,6 +212,31 @@ instance compactSpace : CompactSpace (PrimeSpectrum R) := by
+/-- The prime spectrum of a semiring has discrete Zariski topology iff it is finite and
+all primes are maximal. -/
+theorem discreteTopology_iff_finite_and_isPrime_imp_isMaximal : DiscreteTopology (PrimeSpectrum R) ↔
+    Finite (PrimeSpectrum R) ∧ ∀ I : Ideal R, I.IsPrime → I.IsMaximal :=
```
I really think we should have a `ZeroKrullDimRing` typeclass (not in this PR though).

erdOne reviewed (approved):
Thanks!
maintainer merge

alreadydone commented:
I discovered [TopologicalSpace.Clopens](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Sets/Closeds.html#TopologicalSpace.Clopens) so I switched to use it in [this branch](https://github.com/leanprover-community/mathlib4/compare/DiscreteTopology_PrimeSpectrum...DiscreteTopology_PrimeSpectrum++?expand=1), and added more lemmas around `isIdempotentElemEquivClopens`.
I can merge them now or include them in a subsequent PR, which will show `toLocalizationIsPrime_injective` is a characterization of comm. rings with discrete spectra.

joelriou commented:
> I discovered [TopologicalSpace.Clopens](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Sets/Closeds.html#TopologicalSpace.Clopens) so I switched to use it in [this branch](https://github.com/leanprover-community/mathlib4/compare/DiscreteTopology_PrimeSpectrum...DiscreteTopology_PrimeSpectrum++?expand=1), and added more lemmas around `isIdempotentElemEquivClopens`. I can merge them now or include them in a subsequent PR, which will show `toLocalizationIsPrime_surjective` is a characterization of comm. rings with discrete spectra.

Yes, I think it is better to use `TopologicalSpace.Clopens` when it makes sense.

jcommelin commented on Mathlib/RingTheory/Nilpotent/Basic.lean:
```diff
@@ -91,6 +91,12 @@ lemma IsNilpotent.not_isUnit [Ring R] [Nontrivial R] {x : R} (hx : IsNilpotent x
     ¬ IsUnit x :=
   mt IsUnit.not_isNilpotent (by simpa only [not_not] using hx)
 
+lemma IsIdempotentElem.eq_zero_of_isNilpotent [MonoidWithZero R] {e : R}
```
Could you please add an alias `IsNilpotent.eq_zero_of_isIdempotentElem`?

jcommelin reviewed (commented):
bors d+

alreadydone commented:
Thanks!
bors r+

I think I have a different proof of the main result (isomorphism with product of localizations) that works for semirings, which of course can't use subtraction (as in R/(1-e)R) and the CRT and must make use some form  of the fact that the structure sheaf still works over (comm.) semirings. This will make most lemmas in this PR unnecessary to prove the main result, but I think they're good by themselves.

The result I'll need is essentially some form of [AlgebraicGeometry.StructureSheaf.toBasicOpen_surjective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/AlgebraicGeometry/StructureSheaf.html#AlgebraicGeometry.StructureSheaf.toBasicOpen_surjective) and the preceding lemma, but I don't need to cover a basic open, only the whole spectrum. I have no reason to doubt the proof works over semirings, but the whole file assumes CommRing. Therefore (also to not import this advanced file) I won't directly use these two theorems but will prove a version tailored to my use case, which is fortunately not too complicated.

Side comments: I think the basic opens of idempotents in some (comm.) semirings may not be closed, but I don't know about an example. And it looks like a fun undergraduate project to [compute $\mathrm{Spec}(\mathbb{N})$](https://math.stackexchange.com/a/4224486/12932) in Lean.

alreadydone commented:
Generalization to semirings at #19496

alreadydone commented:
> I think the basic opens of idempotents in some (comm.) semirings may not be closed, but I don't know about an example.

I [found a (class of) examples](https://math.stackexchange.com/a/5004592/12932) when I was about to ask on StackExchange; instead I answered a question open for 7 years, which ask for an idempotent without an orthogonal complement. But [this lemma](https://github.com/leanprover-community/mathlib4/pull/18891/files#diff-60d2f63e563433f6a23c60b9620ae1b1551f8409592cff3326ac1fc5b87e60f9R718-R719) basically shows that clopens correspond to orthogonal pairs of idempotents.

alreadydone commented:
> a subsequent PR, which will show `toLocalizationIsPrime_surjective` is a characterization of comm. rings with discrete spectra.

This is now #19714

alreadydone commented on Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:
```diff
@@ -951,24 +1018,123 @@ lemma PrimeSpectrum.basicOpen_eq_zeroLocus_of_isIdempotentElem
+/-- Clopen subsets in the prime spectrum of a commutative ring are in 1-1 correspondence
+with idempotent elements in the ring. -/
+@[stacks 00EE]
+def isIdempotentElemEquivClopens :
+    {e : R | IsIdempotentElem e} ≃ Clopens (PrimeSpectrum R) :=
```
#20533 generalizes this to semirings, with the LHS replaced by pairs `(e,f)` such that `e + f = 1` and `e * f = 0`.

alreadydone commented:
> And it looks like a fun undergraduate project to [compute](https://math.stackexchange.com/a/4224486/12932) Spec ( N ) in Lean.

The prime spectrum of ℕ is computed in #27065
